File: ov-cloud/app/protocol.py
"""Cloud-side mirror of ov-runner protocol.rs.

Every frame on wss://api.yourservice.io/runner/v1 is a `type`-tagged
envelope. Field names match the Rust serde representation exactly
(snake_case tags, camelCase-free) so the runner and cloud stay in sync.
"""

from __future__ import annotations

from enum import Enum
from typing import Literal, Optional, Union
from uuid import UUID

from pydantic import BaseModel, Field


# ---------- shared payloads ----------

class MemoryConfig(BaseModel):
    mcp_url: str
    user_key: str


class TaskEventKind(str, Enum):
    agent_message_chunk = "agent_message_chunk"
    tool_call = "tool_call"
    tool_call_update = "tool_call_update"
    plan = "plan"
    permission_request = "permission_request"
    error = "error"


class TaskStatus(str, Enum):
    done = "done"
    failed = "failed"
    cancelled = "cancelled"


# ---------- runner -> cloud ----------

class Hello(BaseModel):
    type: Literal["hello"] = "hello"
    token: str
    runner_version: str
    agents: list[str]
    os: str


class Ping(BaseModel):
    type: Literal["ping"] = "ping"


class TaskEvent(BaseModel):
    type: Literal["task_event"] = "task_event"
    task_id: UUID
    kind: TaskEventKind
    text: str
    permission_id: Optional[str] = None


class TaskResult(BaseModel):
    type: Literal["task_result"] = "task_result"
    task_id: UUID
    status: TaskStatus
    session_id: Optional[str] = None
    error: Optional[str] = None


# ---------- cloud -> runner ----------

class AssignTask(BaseModel):
    type: Literal["assign_task"] = "assign_task"
    task_id: UUID
    slack_channel: str
    slack_thread_ts: str
    prompt: str
    agent: str
    cwd: str
    resume_session: Optional[str] = None
    memory: Optional[MemoryConfig] = None


class PermissionDecision(BaseModel):
    type: Literal["permission_decision"] = "permission_decision"
    task_id: UUID
    permission_id: str
    approved: bool


class CancelTask(BaseModel):
    type: Literal["cancel_task"] = "cancel_task"
    task_id: UUID


class Pong(BaseModel):
    type: Literal["pong"] = "pong"


class Error(BaseModel):
    type: Literal["error"] = "error"
    code: str
    message: str


# ---------- envelope union ----------

RunnerMessage = Union[Hello, Ping, TaskEvent, TaskResult]
CloudMessage = Union[AssignTask, PermissionDecision, CancelTask, Pong, Error]


def parse_runner_message(raw: dict) -> RunnerMessage:
    """Dispatch inbound runner frame by its `type` tag."""
    tag = raw.get("type")
    models = {
        "hello": Hello,
        "ping": Ping,
        "task_event": TaskEvent,
        "task_result": TaskResult,
    }
    model = models.get(tag)
    if model is None:
        raise ValueError(f"unknown runner message type: {tag}")
    return model.model_validate(raw)


class CloudOutbound(BaseModel):
    """Serializable outbound frame helper."""
    message: CloudMessage

    def dump(self) -> str:
        return self.message.model_dump_json(exclude_none=True)


def encode(msg: CloudMessage) -> str:
    return msg.model_dump_json(exclude_none=True)

File: runner/crates/farol-core/src/protocol.rs
//! Cloud protocol: messages exchanged between the runner and the SaaS
//! over a single outbound WebSocket (wss://api.farol.team/runner/v1).
//!
//! Direction: the runner always dials out — no open ports on the client.

use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Envelope for every frame on the wire.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum CloudMessage {
    // ---- runner -> cloud ----
    /// First frame after connect: authenticate + announce capabilities.
    Hello(Hello),
    /// Heartbeat, every ~30s.
    Ping,
    /// Streaming chunk of the agent's answer for a task (mapped to a Slack thread).
    TaskEvent(TaskEvent),
    /// Terminal state of a task.
    TaskResult(TaskResult),
    /// User decision on a permission request (relayed FROM Slack via cloud),
    /// answered back to the ACP agent.
    // ---- cloud -> runner ----
    /// New task created by a Slack mention / DM.
    AssignTask(AssignTask),
    /// Slack user pressed Approve/Deny — resolves a pending permission.
    PermissionDecision(PermissionDecision),
    /// Cancel a running task (Stop button / ❌ reaction).
    CancelTask { task_id: Uuid },
    Pong,
    /// Fatal: auth rejected, runner should stop and re-login.
    Error { code: String, message: String },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Hello {
    pub token: String,
    pub runner_version: String,
    /// Agents available locally, e.g. ["claude-code", "gemini"].
    pub agents: Vec<String>,
    pub os: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AssignTask {
    pub task_id: Uuid,
    /// Slack thread this task is bound to (cloud owns the mapping).
    pub slack_channel: String,
    pub slack_thread_ts: String,
    /// Prompt text from the Slack user.
    pub prompt: String,
    /// Which agent to run (cloud may route by user's choice).
    pub agent: String,
    /// Project directory the agent should work in (must be allowlisted locally).
    pub cwd: String,
    /// Optional ACP session id to resume.
    #[serde(default)]
    pub resume_session: Option<String>,
    /// Memory endpoint config injected by the cloud (OpenViking MCP).
    pub memory: Option<MemoryConfig>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MemoryConfig {
    pub mcp_url: String,
    pub user_key: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskEvent {
    pub task_id: Uuid,
    pub kind: TaskEventKind,
    /// Text chunk / tool name / permission description.
    pub text: String,
    /// For permission events: cloud renders Approve/Deny buttons with this id.
    #[serde(default)]
    pub permission_id: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TaskEventKind {
    AgentMessageChunk,
    ToolCall,
    ToolCallUpdate,
    Plan,
    PermissionRequest,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TaskResult {
    pub task_id: Uuid,
    pub status: TaskStatus,
    /// ACP session id — cloud stores it to offer "resume" later.
    pub session_id: Option<String>,
    #[serde(default)]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum TaskStatus {
    Done,
    Failed,
    Cancelled,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PermissionDecision {
    pub task_id: Uuid,
    pub permission_id: String,
    pub approved: bool,
}
